plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/TrinoNessieCatalog.java

```diff
@@ -0,0 +1,450 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class TrinoNessieCatalog
+        extends AbstractTrinoCatalog
+{
+    private final String warehouseLocation;
+    private final NessieIcebergClient nessieClient;
+    // TODO: this is just a workaround for now until Namespaces actually support properties
+    private final Map<String, Map<String, Object>> propertiesByNamespace = new ConcurrentHashMap<>();
+
+    private final Map<SchemaTableName, TableMetadata> tableMetadataCache = new ConcurrentHashMap<>();
+
+    public TrinoNessieCatalog(
+            IcebergTableOperationsProvider tableOperationsProvider,
+            NessieIcebergClient nessieClient,
+            String warehouseLocation,
+            String trinoVersion,
+            boolean useUniqueTableLocation)
+    {
+        super(tableOperationsProvider, trinoVersion, useUniqueTableLocation);
+        this.warehouseLocation = requireNonNull(warehouseLocation, "warehouseLocation is null");
+        this.nessieClient = requireNonNull(nessieClient, "nessieClient is null");
+    }
+
+    @Override
+    public List<String> listNamespaces(ConnectorSession session)
+    {
+        try {
+            return nessieClient.getNessieApi().getMultipleNamespaces()
+                    .refName(nessieClient.getRef().getName())
+                    .get().getNamespaces().stream()
+                    .map(Namespace::name)
+                    .collect(Collectors.toList());
+        }
+        catch (NessieReferenceNotFoundException e) {
+            throw new TrinoException(ICEBERG_CATALOG_ERROR,
+                    "Cannot list Namespaces: ref is no longer valid", e);
+        }
+    }
+
+    @Override
+    public void dropNamespace(ConnectorSession session, String namespace)
+    {
+        try {
+            nessieClient.getRef().checkMutable();
+            nessieClient.getNessieApi().deleteNamespace()
+                    .refName(nessieClient.getRef().getName())
+                    .namespace(namespace)
+                    .delete();
+            propertiesByNamespace.remove(namespace);
+            nessieClient.refreshReference();
+        }
+        catch (NessieNamespaceNotFoundException e) {
+            throw new SchemaNotFoundException(namespace);
+        }
+        catch (NessieReferenceNotFoundException e) {
+            throw new TrinoException(ICEBERG_CATALOG_ERROR,
+                    format("Cannot drop Namespace '%s': ref is no longer valid", namespace),
+                    e);
+        }
+        catch (NessieNamespaceNotEmptyException e) {
+            throw new TrinoException(ICEBERG_CATALOG_ERROR, format(
+                    "Namespace '%s' is not empty. One or more tables exist", namespace), e);
+        }
+    }
+
+    @Override
+    public Map<String, Object> loadNamespaceMetadata(ConnectorSession session, String namespace)
+    {
+        try {
+            nessieClient.getNessieApi().getNamespace()
```

Comment:
Can we encapsulate all the details related to calling `nessieApi` under `nessieClient` ?
Let's have `TrinoNessieCatalog` rely on `nessieClient` for handling the details related to the communication with Nessie.

Reply:
done